adam_core/utils/indexable.py

```diff
@@ -1,174 +1,451 @@
+    the len(instance) would be 3.
+
+    Class To Members Mapping:
+    [0, 0, 0, 1, 1, 1, 2, 2, 2]
+    If, as above, the class to members mapping is sorted (monotonically increasing), then
+    it is converted to an array of slices which allows for faster querying.
+    [slice(0, 2), slice(2, 5), slice(5, 8)]
+
+    Members Index:
+    [0, 1, 2, 3, 4, 5, 6, 7, 8]
+
+    For example, here orbits is a subclass of Indexable and has members that are
+    `~numpy.ndarray`s, `~numpy.ma.core.MaskedArray`s, and `~astropy.time.core.Time`s.
+
+    ```
+    from adam_core.orbits import Orbits
+    from adam_core.backend import PYOORB
+    from astropy.time import Time
+    from astropy import units as u
+
+    # Instantiate an Orbits object with 3 orbits
+    t0 = Time([59000.0], scale="utc", format="mjd")
+    orbits = Orbits.from_horizons(["Eros", "Ceres", "Duende"], t0)
+    len(orbits) == 3
+
+    # If I only wanted one orbit
+    single_orbit = orbits[0] # Orbits with one orbit
+
+    # Lets remove the other ones
+    del orbits[1:]
+
+    # Lets propagate the orbits to 10 new times
+    orbits = Orbits.from_horizons(["Eros", "Ceres", "Duende"], t0)
+    t1 = t0 + np.arange(0, 10) * u.day
+    backend = PYOORB()
+    propagated_orbits = backend.propagate_orbits(orbits, t1)
+
+    # The underlying arrays are now 10 * 3 in length
+    len(propagated_orbits) == 30
+
+    # If we update the index to be on orbit_id
+    # then the length of the class is 3 even though each
+    # member is 10 * 3 in length
+    propagated_orbits.set_index("orbit_id")
+    len(propagated_orbits) == 3
+
+    # Lets delete all of the states for "Duende"
+    del propagated_orbits[2]
+    ```
+    """
+
+    def __init__(self, index_values: Optional[Union[str, npt.ArrayLike]] = None):
+        self.set_index(index_values)
+        return
 
     @property
     def index(self):
-        return self._index
+        """
+        The externally facing index of the class.
+        """
+        return self._class_index
 
-    def set_index(self, index: Union[str, np.ndarray]):
-        if isinstance(index, str):
-            class_index = getattr(self, index)
-            self._index_attribute = index
-            array_index = np.arange(0, len(class_index), dtype=int)
+    @index.setter
+    def index(self, value):
+        self.set_index(value)
 
-        elif isinstance(index, np.ndarray):
-            class_index = index
-            self._index_attribute = None
-            array_index = np.arange(0, len(index), dtype=int)
+    @index.deleter
+    def index(self):
+        self.set_index()
+
+    def _check_member_validity(self):
+        """
+        Scans the class's sliceable members and raises an error if they are
+        not all of the same length.
+
+        Sets
+        ----
+        self._member_length : int
+            The length of the members of the class.
+        self._member_index : `~numpy.ndarray`
+            The index of the members of the class.
```

Comment:
Usually code is cleaner if it returns values rather than setting variables as a side-effect. It will be safer for concurrent use, easier to test, and easier to read. I'd recommend changing this to return the two values. It can still raise an error if internal state is messed up because two members have different sizes.

Reply:
Done.